### PR TITLE
Start binding GEOSContextHandle type

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

**Comment by mthh** [comment]
Oh great ! I really had troubles when I tried to bind `GEOSContext_setNoticeMessageHandler_r` / `GEOSContext_setErrorMessageHandler_r`. I will learn from your example for sure !

**Comment by GuillaumeGomez** [comment]
I just thought that those two functions could have closures will the same lifetime as the object instead of `'static`. That could be interesting.

**Comment by GuillaumeGomez** [comment]
Oh and travis has some weird lazy failures...

File: src/ffi.rs [hunk 1]
```diff
@@ -1,35 +1,41 @@
...
 #[repr(C)]
-struct GEOSCoordSequence { private: [u8; 0]}
+struct GEOSCoordSequence { private: [u8; 0] }
 #[repr(C)]
-struct GEOSGeometry { private: [u8; 0]}
+struct GEOSGeometry { private: [u8; 0] }
+#[repr(C)]
+struct GEOSContextHandle_HS { private: [u8; 0] }
+#[allow(non_camel_case_types)]
+type GEOSContextHandle_t = *mut GEOSContextHandle_HS;
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/ffi.rs]: We could but I'm not a big fan of it. :)

**Comment by GuillaumeGomez** [comment]
Okay, should be ready now!

File: src/ffi.rs [hunk 2]
```diff
@@ -142,22 +150,379 @@ extern "C" {
...
+    /// };
+    /// let point_geom = GGeom::new("POINT (2.5 2.5)").expect("Invalid geometry");
+    /// let wkb_buf = context_handle.geom_to_WKB_buf(&point_geom)
+    ///                             .expect("conversion to WKB failed");
+    /// let new_geom = context_handle.geom_from_WKB_buf(wkb_buf.as_ref())
+    ///                              .expect("conversion to WKB failed");
+    /// assert!(point_geom.equals(&new_geom) == Ok(true));
+    /// ```
+    #[allow(non_snake_case)]
+    pub fn geom_to_WKB_buf(&self, g: &GGeom) -> Option<CVec<u8>> {
```
Comment (GuillaumeGomez) [inline src/ffi.rs]: I just thought: should I return a `String` in here?

File: src/ffi.rs [hunk 3]
```diff
@@ -142,22 +150,379 @@ extern "C" {
...
+    /// };
+    /// let point_geom = GGeom::new("POINT (2.5 2.5)").expect("Invalid geometry");
+    /// let wkb_buf = context_handle.geom_to_WKB_buf(&point_geom)
+    ///                             .expect("conversion to WKB failed");
+    /// let new_geom = context_handle.geom_from_WKB_buf(wkb_buf.as_ref())
+    ///                              .expect("conversion to WKB failed");
+    /// assert!(point_geom.equals(&new_geom) == Ok(true));
+    /// ```
+    #[allow(non_snake_case)]
+    pub fn geom_from_WKB_buf(&self, wkb: &[u8]) -> GeosResult<GGeom> {
```
Comment (GuillaumeGomez) [inline src/ffi.rs]: I just thought: should I take a `String` as parameter in here?

File: src/ffi.rs [hunk 4]
```diff
@@ -142,22 +150,379 @@ extern "C" {
...
+    /// };
+    /// let point_geom = GGeom::new("POINT (2.5 2.5)").expect("Invalid geometry");
+    /// let wkb_buf = context_handle.geom_to_HEX_buf(&point_geom)
+    ///                             .expect("conversion to WKB failed");
+    /// let new_geom = context_handle.geom_from_HEX_buf(wkb_buf.as_ref())
+    ///                              .expect("conversion to WKB failed");
+    /// assert!(point_geom.equals(&new_geom) == Ok(true));
+    /// ```
+    #[allow(non_snake_case)]
+    pub fn geom_from_HEX_buf(&self, hex: &[u8]) -> GeosResult<GGeom> {
```
Comment (GuillaumeGomez) [inline src/ffi.rs]: I just thought: should I take a `String` as parameter in here?

File: src/ffi.rs [hunk 5]
```diff
@@ -142,22 +150,379 @@ extern "C" {
...
+    /// };
+    /// let point_geom = GGeom::new("POINT (2.5 2.5)").expect("Invalid geometry");
+    /// let wkb_buf = context_handle.geom_to_HEX_buf(&point_geom)
+    ///                             .expect("conversion to WKB failed");
+    /// let new_geom = context_handle.geom_from_HEX_buf(wkb_buf.as_ref())
+    ///                              .expect("conversion to WKB failed");
+    /// assert!(point_geom.equals(&new_geom) == Ok(true));
+    /// ```
+    #[allow(non_snake_case)]
+    pub fn geom_to_HEX_buf(&self, g: &GGeom) -> Option<CVec<u8>> {
```
Comment (GuillaumeGomez) [inline src/ffi.rs]: I just thought: should I return a `String` in here?

File: src/ffi.rs (hunk 3 above)
Comment (TeXitoi) [reply src/ffi.rs]: WKB is binary, not utf-8, so `&[u8]` is your only solution.

Comment (TeXitoi) [reply src/ffi.rs]: Why WKB in caps? Why not `geom_from_wkb_buf` as used in the rust convension? https://doc.rust-lang.org/nightly/std/str/fn.from_utf8.html

File: src/ffi.rs [hunk 6]
```diff
@@ -142,22 +150,379 @@ extern "C" {
...
+    ///                              .expect("conversion to WKB failed");
+    /// assert!(point_geom.equals(&new_geom) == Ok(true));
+    /// ```
+    #[allow(non_snake_case)]
+    pub fn geom_to_WKB_buf(&self, g: &GGeom) -> Option<CVec<u8>> {
+        let mut size = 0;
+        unsafe {
+            let ptr = GEOSGeomToWKB_buf_r(self.ptr, g.as_raw(), &mut size);
+            if ptr.is_null() || size == 0 {
+                None
```
Comment (TeXitoi) [inline src/ffi.rs]: If ptr is not null but size=0, you will leak, no?

File: src/ffi.rs [hunk 7]
```diff
@@ -142,22 +150,379 @@ extern "C" {
...
+    ///
+    /// let context_handle = match GContextHandle::init() {
+    ///     Ok(context_handle) => context_handle,
+    ///     Err(e) => panic!("An error occurred: {}", e),
+    /// };
+    /// let point_geom = GGeom::new("POINT (2.5 2.5)").expect("Invalid geometry");
+    /// let wkb_buf = context_handle.geom_to_HEX_buf(&point_geom)
+    ///                             .expect("conversion to WKB failed");
+    /// let new_geom = context_handle.geom_from_HEX_buf(wkb_buf.as_ref())
+    ///                              .expect("conversion to WKB failed");
```
Comment (TeXitoi) [inline src/ffi.rs]: bad expect

File: src/ffi.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/ffi.rs]: I can update the name. :)

File: src/ffi.rs [hunk 8]
```diff
@@ -142,22 +150,379 @@ extern "C" {
...
+impl Drop for GContextHandle {
+    fn drop<'a>(&'a mut self) {
+        unsafe { GEOS_finish_r(self.ptr) };
+
+        let previous_ptr = self.notice_message.replace(::std::ptr::null_mut());
+        if !previous_ptr.is_null() {
+            // We free the previous closure.
+            let _callback: Box<Option<Box<dyn Fn(&str) + 'a>>> =
+                unsafe { Box::from_raw(previous_ptr as *mut _) };
+        }
```
Comment (TeXitoi) [inline src/ffi.rs]: This will mix badly if we use a C lib that also use geos, no?

File: src/ffi.rs (hunk 6 above)
Comment (GuillaumeGomez) [reply src/ffi.rs]: I'm not sure such a case is actually possible...

File: src/ffi.rs (hunk 8 above)
Comment (GuillaumeGomez) [reply src/ffi.rs]: Unless you call this function on the same object but without the binding, it won't. It's not a global function after all.

File: src/ffi.rs (hunk 3 above)
Comment (GuillaumeGomez) [reply src/ffi.rs]: Ah right: I mixed up the name with WKT.

File: src/ffi.rs (hunk 6 above)
Comment (TeXitoi) [reply src/ffi.rs]: as `free` a null ptr is valid, why not just returning a `CVec<u8>` and don't care of the empty case?

Comment (GuillaumeGomez) [reply src/ffi.rs]: Fine by me.

**Comment by GuillaumeGomez** [comment]
Anything else I should change?

File: src/ffi.rs [hunk 9]
```diff
@@ -142,22 +150,379 @@ extern "C" {
...
+    }
+
+    /// Sets WKB output dimensions.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use geos::{GContextHandle, Dimensions};
+    ///
+    /// let context_handle = match GContextHandle::init() {
```
Comment (antoine-de) [inline src/ffi.rs]: by curiosity, why don't you `.expect` here ?

Comment (GuillaumeGomez) [reply src/ffi.rs]: More easy to read but that's just my point of view. I can change it if you prefer?

Comment (antoine-de) [reply src/ffi.rs]: nop, I was just wondering, both are fine

**Comment by GuillaumeGomez** [comment]
Updated.

**Pull request merged**